7_config_merge/src/munge_lake_info.R

```diff
@@ -66,6 +66,45 @@ munge_Kw <- function(out_ind, kw_varying_ind, ...){
   gd_put(out_ind, data_file)
 }
 
+#' @Title Resample incoming hypsography
+#' @decription Resample the A (area) values to a set of elevation intervals
+#' defined based on the depth of the lake
+#' @param site_id id of lake for which hypso is being resampled
+#' @param lake_hypso a dataframe of H (elevation) values and A (area) values
+#' @return a dataframe of the resampled H and A values
+resample_hypso <- function(site_id, lake_hypso) {
+
+  # Determine lake depth
+  lake_depth = diff(range(lake_hypso$H))
+
+  # Use lake depth to determine the elevation interval for the resampled hypsography
+  interval <- case_when(
+    lake_depth >= 8 ~ 1,
+    lake_depth >= 5 & lake_depth < 8 ~ 0.5,
+    TRUE ~ 0.25
+  )
+
+  # Resample hypso to defined interval, using approx() to resample the
+  # radii rather than the raw area, per Lindsay's approach in lake-temperature-out:
+  # https://github.com/USGS-R/lake-temperature-out/blob/main/2_process/src/calculate_toha.R#L302-L313
+  lake_hypso$radii = sqrt(lake_hypso$A/pi)
+  hypso_resampled <- bind_rows(
+    # Add an additional row for the minimum (initial, deepest) raw H value so we retain the original area at the bottom of the lake
+    tibble(new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=min(lake_hypso$H), rule=2)$y, H=min(lake_hypso$H)),
+    # Between min and max elevation (non-inclusive), interpolate based on specified interval
+    tibble(H=seq(floor((min(lake_hypso$H)+interval)/interval)*interval, floor((max(lake_hypso$H)-interval)/interval)*interval, by=interval),
+           new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=H, rule=2)$y),
+    # Add an additional row for the maximum (final, surface) raw H value so we retain the original surface area
+    tibble(new_radii=approx(lake_hypso$H, lake_hypso$radii, xout=max(lake_hypso$H), rule=2)$y, H=max(lake_hypso$H))) %>%
+    mutate(A = pi * new_radii^2) %>%
```

Comment:
There is a lot of repeated work happening here, so I think there is a way to simplify by using the fxn I shared as a little helper/utility. I think you can hide the `radii` stuff and focus this code on defining the depths that you want to have returned.

Use the fxn from `lake-temperature-out` with a slight adjustment to return a data.frame with A and H, not depths/areas:

```r
resample_hypso_util <- function(hypso, new_H) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$A / pi)
  new_radii <- approx(hypso$H, hypso$radii, xout=new_H, rule = 2)$y
  
  # Now calculate area from new radii
  new_A <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(data.frame(H = new_H, A = new_A))
}
```

So start by just defining the new H's using your approach above but making a single vector of all these new Hs. Then, just pass those into the little helper fxn.

```r
new_interval_elevations <- c(
  min(lake_hypso$H),
  seq(from = floor((min(lake_hypso$H)+interval)/interval)*interval, 
      to = floor((max(lake_hypso$H)-interval)/interval)*interval, 
      by=interval),
  max(lake_hypso$H)
)

resample_hypso_util(lake_hypso, new_interval_elevations)
```

When I run a test of this, I get the same output as what you shared in the updated NML rds file for both site `nhdhr_105954667` and `nhdhr_106373442`

```r
# Load util fxn
resample_hypso_util <- function(hypso, new_H) {
  
  # `rule = 2`: repeats top or bottom known hypso for any depths outside of known hypso
  hypso$radii <- sqrt(hypso$A / pi)
  new_radii <- approx(hypso$H, hypso$radii, xout=new_H, rule = 2)$y
  
  # Now calculate area from new radii
  new_A <- pi * new_radii^2
  
  # Create hypso at these new depths
  return(data.frame(H = new_H, A = new_A))
}

# Morphometry before resampling for nhdhr_105954667
site1_morph <- list(
  H = c(446.6848, 447.904, 449.428, 450.952, 452.476, 454), 
  A = c(0, 21045.3337177386, 83253.6667468361, 259155.913693587, 
        383881.802459821, 454739.276528491)
)

# Morphometry before resampling for nhdhr_106373442
site2_morph <- list(
  H = c(398.5888, 398.8936, 399.1984, 399.5032, 399.808, 400.1128, 
        400.4176, 400.7224, 401.0272, 401.332, 401.6368, 401.9416, 
        402.2464, 402.5512, 402.856, 403.1608, 403.4656, 403.7704, 
        404.0752, 404.38, 404.6848, 404.9896, 405.2944, 405.5992, 
        405.904, 406.2088, 406.5136, 406.8184, 407.1232, 407.428, 
        407.7328, 408.0376, 408.3424, 408.6472, 408.952, 409.2568, 
        409.5616, 409.8664, 410.1712, 410.476, 410.7808, 411.0856, 
        411.3904, 411.6952, 412), 
  A = c(3699.73951147273, 39297.2331894265, 
        69870.0806389613, 99068.024891665, 159888.742671484, 191711.502118137, 
        203660.660810597, 213909.939186974, 224209.214043236, 235183.441377942, 
        247132.600070401, 260706.644359115, 278130.417598957, 306828.397052813, 
        350950.290551119, 386247.805349764, 417995.570076591, 447243.51080918, 
        475566.5166639, 504189.501397929, 533362.44741069, 563735.308940686, 
        596932.971584171, 662203.376073734, 709375.054845011, 743797.631245673, 
        775395.406532845, 869663.769355573, 929834.532896889, 964957.060015938, 
        986480.54460633, 1005079.23512346, 1022952.97668227, 1041351.68127986, 
        1062700.17819066, 1085323.58533851, 1104997.20017317, 1124445.83084834, 
        1143644.47912409, 1163868.05523748, 1184116.62959081, 1205665.11242115, 
        1228963.47204745, 1256686.52014355, 1300033.46820364)
)

# CHANGE ME to `site1_morph` to test the other site outputs
lake_hypso <- site2_morph 

lake_depth <- diff(range(lake_hypso$H))
interval <- dplyr::case_when(
    lake_depth >= 8 ~ 1,
    lake_depth >= 5 & lake_depth < 8 ~ 0.5,
    TRUE ~ 0.25
)

new_interval_elevations <- c(
  min(lake_hypso$H),
  seq(from = floor((min(lake_hypso$H)+interval)/interval)*interval, 
      to = floor((max(lake_hypso$H)-interval)/interval)*interval, 
      by=interval),
  max(lake_hypso$H)
)

resample_hypso_util(lake_hypso, new_interval_elevations)
```

```r
# Resampling results for site1 (nhdhr_105954667)
          H          A
1  446.6848      0.000
2  447.0000   1406.625
3  447.5000   9408.800
4  448.0000  23749.088
5  448.5000  40472.033
6  449.0000  61626.004
7  449.5000  89374.786
8  450.0000 137871.595
9  450.5000 196838.747
10 451.0000 262711.769
11 451.5000 301192.625
12 452.0000 342302.481
13 452.5000 384951.186
14 453.0000 407568.299
15 453.5000 430830.996
16 454.0000 454739.277

# Resampling results for site2 (nhdhr_106373442)
          H          A
1  398.5888    3699.74
2  399.0000   48977.03
3  400.0000  179598.17
4  401.0000  223280.28
5  402.0000  264001.41
6  403.0000  367415.54
7  404.0000  468497.89
8  405.0000  564852.39
9  406.0000  720128.83
10 407.0000  905271.21
11 408.0000 1002775.52
12 409.0000 1066247.11
13 410.0000 1132840.98
14 411.0000 1199593.82
15 412.0000 1300033.47
```

Reply:
Alright - suggested edits to reorganize and streamline this code have been made